### PR TITLE
BMCDiscovery: Initial commit

### PR DESCRIPTION
**Pull request by PSayenka**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Initial BMC Discovery Integration PR

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [x ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x ] No

## Must have
- [x ] Tests
- [ x] Documentation

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -3,2309 +3,13 @@
   "version": "1.0.0",
```
Comment (yaakovi) [inline package-lock.json:3]: restore this file

File: Packs/BMCDiscovery/README.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+BMC Discovery is a SaaS-based, cloud-native discovery and dependency modeling system that provides instant visibility into hardware, software, and service dependencies across multi-cloud, hybrid, and on-premises environments.
```
Comment (yaakovi) [inline Packs/BMCDiscovery/README.md:1]: * Use line 1 as the pack description.
* Move line 2 to the integration README.md file.

File: Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+  required: true
+  type: 0
+- display: ""
+  displaypassword: API Token
+  name: api_token
+  type: 9
+  required: true
+  hiddenusername: true
+- defaultvalue: 'false'
+  display: Use system proxy settings
```
Comment (yaakovi) [inline Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml:17]:
```suggestion
- display: Use system proxy settings
```

File: Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+  name: proxy
+  required: false
+  type: 8
+- defaultvalue: 'false'
+  display: Trust any certificate (not secure)
```
Comment (yaakovi) [inline Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml:22]:
```suggestion
- display: Trust any certificate (not secure)
```

File: Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+  name: insecure
+  required: false
+  type: 8
+description: BMC Discovery Integration
```
Comment (yaakovi) [inline Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml:26]: use the description you have in the README

File: Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+  #  description: Get discovery status
+  #  execution: false
+  #  name: bmc-discovery-process-status
+  #  outputs:
+  #  - contextPath: BmcDiscovery.Process.status
+   #   description: Displays BMC Discovery overall status
+  #    type: String
+  #  - contextPath: BmcDiscovery.Process.running
+  #    description: Displays if any active scans are running
+  #    type: Boolean
```
Comment (yaakovi) [inline Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml:42]: Remove if not needed

File: Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.py [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+        data = {
+            'query': query
+        }
+        return self._http_request(method=method, url_suffix=url_suffix, resp_type='json',
+                                  params=params, json_data=data, ok_codes=(200,))
+
+    def discovery_scan_create(self, **args):
+        url_suffix = '/discovery/runs'
+        label = args.get('label', '')
+        ranges = list(args.get('ranges', '').split(','))
```
Comment (yaakovi) [inline Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.py:91]: You can use `argToList` instead.

File: Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.py [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+        status = demisto.get(response, 'status')
+        if not status:
+            return f'Unexpected result from the service: status={status}'
+        return 'ok'
+    except Exception as e:
+        exception_text = str(e).lower()
+        if 'forbidden' in exception_text or 'authorization' in exception_text:
+            return 'Authorization Error: make sure API Key is correctly set'
+        else:
+            raise e
```
Comment (yaakovi) [inline Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.py:147]:
```suggestion
            raise
```

File: Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+    count = 0
+    for item in response:
+        results = item['results']
+        count += item.get('count', 0)
+        for result in results:
+            keys = list(result.keys())
+            for k in keys:
+                if not result[k]:
+                    del result[k]
+            output['data'].append(result)
```
Comment (yaakovi) [inline Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.py:189]: you can use `remove_nulls_from_dictionary`.

File: Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.py [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+        elif command == 'bmc-discovery-scan-stop':
+            return_results(discovery_scan_stop_command(client, **args))
+        elif command == 'bmc-discovery-scan-summary':
+            return_results(discovery_scan_summary_command(client, **args))
+        elif command == 'bmc-discovery-scan-results-list':
+            return_results(discovery_scan_results_list_command(client, **args))
+        else:
+            raise NotImplementedError(f"command {command} is not implemented.")
+    except Exception as e:
+        demisto.error(traceback.format_exc())
```
Comment (yaakovi) [inline Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.py:366]:
```suggestion
        demisto.error(fix_traceback_line_numbers(traceback.format_exc()))
```

**Review by yaakovi — COMMENTED** [review]
very nice work!
see my comments

File: Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+  dockerimage: demisto/python3:3.10.4.27798
+  feed: false
+  isfetch: false
+  longRunning: false
+  longRunningPort: false
+  runonce: false
+  script: '-'
+  subtype: python3
+  type: python
+fromversion: 6.0.0
```
Comment (yaakovi) [inline Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml:310]:
```suggestion
fromversion: 6.1.0
```

File: Packs/BMCDiscovery/TestPlaybooks/BMC_Discovery_-_Test.yml [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+        "height": 1155,
+        "width": 1350,
+        "x": 50,
+        "y": 50
+      }
+    }
+  }
+inputs: []
+outputs: []
+fromversion: 6.0.0
```
Comment (yaakovi) [inline Packs/BMCDiscovery/TestPlaybooks/BMC_Discovery_-_Test.yml:373]:
```suggestion
fromversion: 6.1.0
```

**Review by yaakovi — CHANGES_REQUESTED** [review]

File: package-lock.json (hunk 1 above)
Comment (PSayenka) [reply package-lock.json:3]: Thank you! Will do.

File: Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+  required: false
+  type: 8
+- display: Trust any certificate (not secure)
+  name: insecure
+  required: false
+  type: 8
+description: BMC Discovery is a SaaS-based, cloud-native discovery and dependency modeling system that provides instant visibility into hardware, software, and service dependencies across multi-cloud, hybrid, and on-premises environments.
+detaileddescription: ''
+display: BMC Discovery
+name: bmcdiscovery
```
Comment (yaakovi) [inline Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml:27]:
```suggestion
name: BMC Discovery
```

File: Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml [hunk 14]
```diff
@@ -0,0 +1,317 @@
+category: Endpoint
+commonfields:
+  id: bmcdiscovery
```
Comment (yaakovi) [inline Packs/BMCDiscovery/Integrations/BMCDiscovery/BMCDiscovery.yml:3]:
```suggestion
  id: BMC Discovery
```

**Review by yaakovi — APPROVED** [review]

**Pull request merged**